tools/kolla-images.py

```
@@ -0,0 +1,365 @@
+            prefix = "stable/"
+            assert value.startswith(prefix)
+            return value[len(prefix):]
+    raise Exception("Failed to determine OpenStack release")
+
+
+def validate(kolla_image_tags: KollaImageTags):
+    """Validate the kolla_image_tags variable."""
+    tag_var_re = re.compile(r"^[a-z0-9_]+$")
+    openstack_release = get_openstack_release()
+    tag_res = {
+        base_distro: re.compile(f"^{openstack_release}-{base_distro}-[\d]{{8}}T[\d]{{6}}$")
+        for base_distro in SUPPORTED_BASE_DISTROS
+    }
+    errors = []
+    if "openstack" not in kolla_image_tags:
+        errors.append("Missing default openstack tag")
+    for tag_var, base_distros in kolla_image_tags.items():
+        if not tag_var_re.match(tag_var):
+            errors.append(f"Key {tag_var} does not match expected pattern. It should match {tag_var_re.pattern}")
+        for base_distro, tag in base_distros.items():
+            if base_distro not in SUPPORTED_BASE_DISTROS:
+                errors.append(f"{tag_var}: base distro {base_distro} not supported. Options: {SUPPORTED_BASE_DISTROS}")
+                continue
+            if not tag_res[base_distro].match(tag):
+                errors.append(f"{tag_var}: {base_distro}: tag {tag} does not match expected pattern. It should match {tag_res[base_distro].pattern}")
+    if errors:
+        print("Errors in kolla_image_tags variable:")
+        for error in errors:
+            print(error)
+        sys.exit(1)
+
+
+def check_tags(base_distros: List[str], kolla_image_tags: KollaImageTags, registry: str, namespace: str):
+    """Check whether expected tags are present in container image registry."""
+    image_tags = get_tags(base_distros, kolla_image_tags)
+
+    missing = {}
+    for image, tags in image_tags.items():
+        for _ in range(3):
+            try:
+                output = subprocess.check_output(f"skopeo list-tags docker://{registry}/{namespace}/{image}", shell=True)
```

Is skopeo expected to be present?

---

Yes. Currently this is just used for local debugging. I could add a `type skopeo` to fail early if it's not installed.

---

Done